### PR TITLE
Sg2/referable interpretation fix

### PR DESCRIPTION
**Pull request by elizabeth-legros**
Changing interpreter to support multiple context nodes
Changing interpreter to create remap blocks for non-context node outputs.
Adding support for reference nodes to interpreter

### CODE REVIEW

File: com.unity.shadergraph/Editor/GraphDeltaRegistry/Generation/Interpreter.cs [hunk 1]
```diff
@@ -263,7 +344,14 @@ private static bool FunctionsAreEqual(ShaderFunction a, ShaderFunction b)
                         if (connectedPort != null) // connected input port-
                         {
                             var connectedNode = connectedPort.GetNode();
-                            argument = $"SYNTAX_{connectedNode.ID.LocalPath}_{connectedPort.ID.LocalPath}";
+                            if (connectedNode.HasMetadata("_contextDescriptor"))
+                            {
+                                argument = $"In.{connectedPort.ID.LocalPath.Replace("out_", "")}";
```
Comment (elizabeth-legros) [inline com.unity.shadergraph/Editor/GraphDeltaRegistry/Generation/Interpreter.cs:349]: This is gross and should probably be supported better

File: com.unity.shadergraph/Editor/GraphDeltaRegistry/Registry/ReferenceNodeBuilder.cs [hunk 2]
```diff
@@ -31,9 +34,9 @@ public ShaderFunction GetShaderFunction(NodeHandler node, ShaderContainer contai
             var shaderType = registry.GetShaderType(field, container);
 
             var shaderFunctionBuilder = new ShaderFunction.Builder(container, $"refpass_{shaderType.Name}");
-            shaderFunctionBuilder.AddInput(shaderType, "In");
-            shaderFunctionBuilder.AddOutput(shaderType, "Out");
-            shaderFunctionBuilder.AddLine("Out = In;");
+            shaderFunctionBuilder.AddInput(shaderType, "Input");
```
Comment (elizabeth-legros) [inline com.unity.shadergraph/Editor/GraphDeltaRegistry/Registry/ReferenceNodeBuilder.cs:37]: Rename here to match ShaderFunction expectations in Interpreter

File: com.unity.shadergraph/Tests/HeadlessPreview/PreviewTests.cs [hunk 3]
```diff
@@ -526,7 +526,7 @@ public void NodePreview_GetFuncAndBlockShaderCode()
...
+            string expectedBlock = "void PassThrough(out float1 Out, float1 In)\n{\nOut = In.x;\n}\nShaderGraphBlock_PassNodeInstBlock::ShaderGraphBlock_PassNodeInstOutput SYNTAX_PassNodeInstMain(ShaderGraphBlock_PassNodeInstBlock::ShaderGraphBlock_PassNodeInstInput In)\n{\nfloat1 SYNTAX_PassNodeInst_Out = float1(0);\nPassThrough(SYNTAX_PassNodeInst_Out, float1(1));\nShaderGraphBlock_PassNodeInstBlock::ShaderGraphBlock_PassNodeInstOutput output;\noutput.Out = SYNTAX_PassNodeInst_Out;\nreturn output;\n}\nShaderGraphBlock_PassNodeInst_REMAPBlock::ShaderGraphBlock_PassNodeInst_REMAPOutput SYNTAX_ShaderGraphBlock_PassNodeInst_REMAPMain(ShaderGraphBlock_PassNodeInst_REMAPBlock::ShaderGraphBlock_PassNodeInst_REMAPInput inputs)\n{\nShaderGraphBlock_PassNodeInst_REMAPBlock::ShaderGraphBlock_PassNodeInst_REMAPOutput output;\noutput.BaseColor = float3(inputs.Out, inputs.Out, inputs.Out);\nreturn output;\n}\n";
```
Comment (elizabeth-legros) [inline com.unity.shadergraph/Tests/HeadlessPreview/PreviewTests.cs:529]: Updating expected code to take into account block remap

**Review by esmelusina — APPROVED** [review]

File: com.unity.shadergraph/Tests/HeadlessPreview/PreviewTests.cs (hunk 3 above)
Comment (esmelusina) [reply com.unity.shadergraph/Tests/HeadlessPreview/PreviewTests.cs:529]: Does this pass?

Comment (esmelusina) [reply com.unity.shadergraph/Tests/HeadlessPreview/PreviewTests.cs:529]: My bad- I had the wrong version of the branch.

**Pull request merged**